runtime/onert/core/src/loader/train/CheckpointLoader.cc

```diff
@@ -32,6 +32,48 @@ using namespace train;
 using namespace checkpoint;
 using namespace exec;
 
+struct DataBufferPair
+{
+  uint32_t offset;
+  uint32_t size;
+};
+
+class DataBuffer
+{
+public:
+  DataBuffer() = default;
+  DataBuffer(uint32_t size) { setSize(size); }
+
+  void setSize(uint32_t size)
+  {
+    _offset.resize(size);
+    _size.resize(size);
+  }
+
+  char *getOffsetBuf() { return reinterpret_cast<char *>(_offset.data()); }
+
+  // This function should be called after loading the _offset buffer.
+  void calculateSize(uint32_t next_start_offset)
+  {
+    assert(_offset.size() == _size.size());
+    for (size_t i = 0; i < _offset.size() - 1; ++i)
+      _size[i] = _offset[i + 1] - _offset[i];
+    _size.back() = next_start_offset - _offset.back();
+  }
+
+  // offset, size
+  DataBufferPair operator[](uint32_t i) const
+  {
+    assert(_offset.size() == _size.size());
+    assert(i <= _offset.size());
+    return DataBufferPair{_offset[i], _size[i]};
```

Comment:
Unlike [std::map::operator[]](https://en.cppreference.com/w/cpp/container/map/operator_at), [std::vector::operator[]](https://en.cppreference.com/w/cpp/container/vector/operator_at) never inserts a new element into the container.
```suggestion
    assert(i < _offset.size());
    return DataBufferPair{_offset[i], _size[i]};
```